File: DemoObjC/SJPageViewController/Demo/SJViewController4.m
//
//  SJViewController0.m
//  SJPageViewController_Example
//
//  Created by BlueDancer on 2020/2/8.
//  Copyright © 2020 changsanjiang@gmail.com. All rights reserved.
//

#import "SJViewController4.h"
#import <SJPageViewController/SJPageViewController.h>
#import <Masonry/Masonry.h>
#import "SJDemoTableViewController.h"


@interface SJNavigationBar : UIView
@property (nonatomic, strong, readonly) UIView *contentView;
@property (nonatomic, strong, readonly) UIButton *backButton;
@end

@implementation SJNavigationBar
- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if ( self ) {
        _contentView = [UIView.alloc initWithFrame:CGRectZero];
        _contentView.backgroundColor = UIColor.orangeColor;
        _contentView.layer.borderWidth = 0.5;
        _contentView.layer.borderColor = UIColor.blackColor.CGColor;
        [self addSubview:_contentView];
        
        _backButton = [UIButton buttonWithType:UIButtonTypeSystem];
        [_backButton setTitle:@"  <--返回按钮" forState:UIControlStateNormal];
        [self addSubview:_backButton];
        
        [_contentView mas_makeConstraints:^(MASConstraintMaker *make) {
            make.edges.offset(0);
        }];
        
        [_backButton mas_makeConstraints:^(MASConstraintMaker *make) {
            if (@available(iOS 11.0, *)) {
                make.top.equalTo(self.mas_safeAreaLayoutGuideTop);
            } else {
                make.top.offset(20);
            }
            make.left.bottom.offset(0);
            make.height.offset(44);
        }];
    }
    return self;
}
@end

@interface SJViewController4 ()<SJPageViewControllerDelegate, SJPageViewControllerDataSource>
@property (nonatomic, strong) SJNavigationBar *navBar;
@property (nonatomic, strong) SJPageViewController *pageViewController;
@end

@implementation SJViewController4

- (void)viewDidLoad {
    [super viewDidLoad];
    [self _setupViews];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    [self.navigationController setNavigationBarHidden:YES animated:YES];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    [self.navigationController setNavigationBarHidden:NO animated:YES];
}

- (void)dealloc {
#ifdef DEBUG
    NSLog(@"%d \t %s", (int)__LINE__, __func__);
#endif
}

- (void)backButtonWasTapped {
    [self.navigationController popViewControllerAnimated:YES];
}

#pragma mark -

- (void)_setupViews {
    self.view.backgroundColor = UIColor.blackColor;
    
    _navBar = [SJNavigationBar.alloc initWithFrame:CGRectZero];
    _navBar.contentView.alpha = 0;
    [_navBar.backButton addTarget:self action:@selector(backButtonWasTapped) forControlEvents:UIControlEventTouchUpInside];
    
    _pageViewController = [SJPageViewController pageViewControllerWithOptions:@{SJPageViewControllerOptionInterPageSpacingKey:@(3)}];
    _pageViewController.dataSource = self;
    _pageViewController.delegate = self;
    [self addChildViewController:_pageViewController];
    [self.view addSubview:_pageViewController.view];
    [self.view addSubview:_navBar];
    
    [_navBar mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.left.right.offset(0);
    }];
    
    [_pageViewController.view mas_makeConstraints:^(MASConstraintMaker *make) {
        make.edges.offset(0);
    }];
}

- (NSUInteger)numberOfViewControllersInPageViewController:(SJPageViewController *)pageViewController {
    return 3;
}

- (UIViewController *)pageViewController:(SJPageViewController *)pageViewController viewControllerAtIndex:(NSInteger)index {
    return SJDemoTableViewController.new;
}

///
/// 顶部下拉时, headerView 固定在顶部
///
- (SJPageViewControllerHeaderMode)modeForHeaderWithPageViewController:(SJPageViewController *)pageViewController {
    return SJPageViewControllerHeaderModePinnedToTop;
}

///
/// 头部视图
///
- (nullable __kindof UIView *)viewForHeaderInPageViewController:(SJPageViewController *)pageViewController {
    UIView *headerView = [UIView.alloc initWithFrame:CGRectMake(0, 0, self.view.bounds.size.width, 375)];
    headerView.backgroundColor = UIColor.purpleColor;
    
    UILabel *channelView = [UILabel.alloc initWithFrame:CGRectZero];
    channelView.text = @"左右滑动切换vc";
    channelView.backgroundColor = UIColor.greenColor;
    [headerView addSubview:channelView];
    [channelView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.bottom.right.offset(0);
        make.height.offset(44);
    }];
    
    return headerView;
}
 
///
/// 在顶部悬浮保留的高度
///
- (CGFloat)heightForHeaderPinToVisibleBoundsWithPageViewController:(SJPageViewController *)pageViewController {
    return 44 + _navBar.bounds.size.height;
}

#pragma mark -

///
/// HeaderView 可见范围发生改变的回调
///
- (void)pageViewController:(SJPageViewController *)pageViewController headerViewVisibleRectDidChange:(CGRect)visibleRect {
    /// headerView的高度
    CGFloat headerViewHeight = pageViewController.heightForHeaderBounds;
    /// 在顶部固定时的高度
    CGFloat pinnedHeight = pageViewController.heightForHeaderPinToVisibleBounds;
    /// 设置导航栏透明度
    CGFloat alpha = 1 - (visibleRect.size.height - pinnedHeight) / (headerViewHeight - pinnedHeight);
    _navBar.contentView.alpha = alpha;
}
@end

File: DemoSwift/SJPageViewControllerswift/Demo/SJViewController4.swift
//
//  SJViewController4.swift
//  SJPageViewControllerswift_Example
//
//  Created by BlueDancer on 2020/2/9.
//  Copyright © 2020 CocoaPods. All rights reserved.
//

import UIKit
import SJPageViewController

class SJNavigationBar: UIView {
    var contentView: UIView
    var backButton: UIButton
    
    override init(frame: CGRect) {
        contentView = UIView.init()
        contentView.backgroundColor = .orange
        contentView.layer.borderWidth = 0.5
        contentView.layer.borderColor = UIColor.black.cgColor
        
        backButton = UIButton.init(type: .system)
        backButton.setTitle("  <--返回按钮", for: .normal)
        super.init(frame: frame)
        self.addSubview(contentView)
        self.addSubview(backButton)
        
        contentView.translatesAutoresizingMaskIntoConstraints = false
        contentView.topAnchor.constraint(equalTo: self.topAnchor).isActive = true
        contentView.leftAnchor.constraint(equalTo: self.leftAnchor).isActive = true
        contentView.bottomAnchor.constraint(equalTo: self.bottomAnchor).isActive = true
        contentView.rightAnchor.constraint(equalTo: self.rightAnchor).isActive = true
        
        backButton.translatesAutoresizingMaskIntoConstraints = false
        if #available(iOS 11.0, *) {
            backButton.topAnchor.constraint(equalTo: self.safeAreaLayoutGuide.topAnchor).isActive = true
        } else {
            let topCons = backButton.topAnchor.constraint(equalTo: self.topAnchor)
            topCons.constant = 20
            topCons.isActive = true
        }
        backButton.leftAnchor.constraint(equalTo: self.leftAnchor).isActive = true
        backButton.bottomAnchor.constraint(equalTo: self.bottomAnchor).isActive = true
        backButton.heightAnchor.constraint(equalToConstant: 44).isActive = true
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}


class SJViewController4: UIViewController {

    var pageViewController: SJPageViewController!
    var navBar: SJNavigationBar!
    
    override func viewDidLoad() {
        super.viewDidLoad() 
        self.view.backgroundColor = .black
        
        navBar = SJNavigationBar.init(frame: .zero)
        navBar.contentView.alpha = 0
        navBar.backButton.addTarget(self, action: #selector(backButtonWasTapped), for: .touchUpInside)
        
        pageViewController = SJPageViewController.init(options: [.interPageSpacing:CGFloat(3)], dataSource: self, delegate: self)
        pageViewController.bounces = false;
        self.addChild(pageViewController)
        self.view.addSubview(pageViewController.view)
        self.view.addSubview(navBar)
        
        navBar.translatesAutoresizingMaskIntoConstraints = false
        navBar.topAnchor.constraint(equalTo: self.view.topAnchor).isActive = true
        navBar.leftAnchor.constraint(equalTo: self.view.leftAnchor).isActive = true
        navBar.rightAnchor.constraint(equalTo: self.view.rightAnchor).isActive = true

        pageViewController.view.translatesAutoresizingMaskIntoConstraints = false
        pageViewController.view.topAnchor.constraint(equalTo: self.view.topAnchor).isActive = true
        pageViewController.view.leftAnchor.constraint(equalTo: self.view.leftAnchor).isActive = true
        pageViewController.view.bottomAnchor.constraint(equalTo: self.view.bottomAnchor).isActive = true
        pageViewController.view.rightAnchor.constraint(equalTo: self.view.rightAnchor).isActive = true
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        self.navigationController?.setNavigationBarHidden(true, animated: true)
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        self.navigationController?.setNavigationBarHidden(false, animated: true)
    }
    
    @objc func backButtonWasTapped() {
        self.navigationController?.popViewController(animated: true)
    }

    override func viewWillLayoutSubviews() {
        super.viewWillLayoutSubviews()
        pageViewController.view.frame = self.view.bounds
    }
}

extension SJViewController4: SJPageViewControllerDataSource {
    func numberOfViewControllers(in pageViewController: SJPageViewController) -> Int {
        return 3
    }
    
    func pageViewController(_ pageViewController: SJPageViewController, viewControllerAt index: Int) -> UIViewController {
        return SJDemoTableViewController.init()
    }
    
    func viewForHeader(in pageViewController: SJPageViewController) -> UIView? {
        let headerView = UIView.init(frame: .init(x: 0, y: 0, width: self.view.bounds.size.width, height: 300))
        headerView.backgroundColor = .purple
        
        let label = UILabel.init(frame: .init(x: 0, y: 300 - 44, width: self.view.bounds.size.width, height: 44))
        label.text = "左右滑动切换vc"
        label.backgroundColor = .green
        headerView.addSubview(label)
        return headerView
    }
    
    func heightForHeaderBounds(with pageViewController: SJPageViewController) -> CGFloat {
        return 300
    }
    
    func heightForHeaderPinToVisibleBounds(with pageViewController: SJPageViewController) -> CGFloat {
        return 44 + navBar.bounds.height
    }
    
    func modeForHeader(with pageViewController: SJPageViewController) -> SJPageViewController.HeaderMode {
        return .pinnedToTop
    }
}

extension SJViewController4: SJPageViewControllerDelegate {
    func pageViewController(_ pageViewController: SJPageViewController, headerViewScrollProgressDidChange progress: CGFloat) {
        navBar.contentView.alpha = progress
    }
}
